frontend/src/metabase/visualizations/components/Visualization/Visualization.jsx

```diff
@@ -72,39 +72,87 @@ const mapStateToProps = state => ({
 
 const SMALL_CARD_WIDTH_THRESHOLD = 150;
 
-class Visualization extends PureComponent {
-  state = {
-    getHref: undefined,
-    hovered: null,
-    clicked: null,
-    error: null,
-    genericError: null,
-    warnings: [],
-    series: null,
-    visualization: null,
-    computedSettings: {},
+const isLoading = series => {
+  return !(
+    series &&
+    series.length > 0 &&
+    _.every(
+      series,
+      s => s.data || _.isObject(s.card.visualization_settings.virtual_card),
+    )
+  );
+};
+
+const deriveStateFromProps = props => {
+  const transformed = props.rawSeries
+    ? getVisualizationTransformed(extractRemappings(props.rawSeries))
+    : null;
+
+  const series = transformed?.series;
+
+  const computedSettings = !isLoading(series)
+    ? getComputedSettingsForSeries(series)
+    : {};
+
+  return {
+    series,
+    computedSettings,
+    visualization: transformed?.visualization,
   };
+};
+
+class Visualization extends PureComponent {
+  constructor(props) {
+    super(props);
 
-  UNSAFE_componentWillMount() {
-    this.transform(this.props);
+    this.state = {
+      getHref: undefined,
+      hovered: null,
+      clicked: null,
+      error: null,
+      genericError: null,
+      warnings: [],
+      series: null,
+      visualization: null,
+      computedSettings: {},
+      ...deriveStateFromProps(props),
+    };
   }
 
-  UNSAFE_componentWillReceiveProps(newProps) {
+  static getDerivedStateFromProps(props, state) {
+    // When these props has changed, we need to re-derive the state.
+    // getDerivedStateFromProps does not have access to the last props, so
     if (
-      !isSameSeries(newProps.rawSeries, this.props.rawSeries) ||
-      !equals(newProps.settings, this.props.settings) ||
-      !equals(newProps.timelineEvents, this.props.timelineEvents) ||
+      !isSameSeries(props.rawSeries, state._lastProps?.rawSeries) ||
+      !equals(props.settings, state._lastProps?.settings) ||
+      !equals(props.timelineEvents, state._lastProps?.timelineEvents) ||
       !equals(
-        newProps.selectedTimelineEventIds,
-        this.props.selectedTimelineEventIds,
+        props.selectedTimelineEventIds,
+        state._lastProps?.selectedTimelineEventIds,
       )
     ) {
-      this.transform(newProps);
+      return {
+        ...state,
+        ...deriveStateFromProps(props),
+
+        // Reset the state to its initial values when these props have changed
+        hovered: null,
+        error: null,
+        genericError: null,
+        warnings: [],
+
+        // Store last properties to compare with the next call
+        _lastProps: _.pick(props, [
+          "rawSeries",
+          "settings",
+          "timelineEvents",
+          "selectedTimelineEventIds",
+        ]),
```

Comment:
Annoyingly enough `getDerivedStateFromProps` does not give us access to the previous props. The only way to use it is to store these ourselves in the state and compare it. Ouch.